### ISSUE TITLE
Add chapter marks

### ISSUE THREAD

**Issue by PaulWoitaschek** [PaulWoitaschek/MaterialAudiobookPlayer#205]
1. Find a way to read the chapter marks from the audiofiles. **Need suggestions how to do this**
2. Replace chapter name with a HashMap of <String, Integer> (Title, StartTime).
3. Fetch the chapter metadata in BookAdder.java and add them to the new chapters
4. Show chapters instead of files in BookPlay
5. Let MediaPlayerControllers previous / next functions redirect to a seekTo if there are chapters in the file.

**Comment by kgraefe** [comment]
I think [AntennaPod](https://github.com/AntennaPod) supports this. [java/de/danoeh/antennapod/core/util/id3reader/ChapterReader.java](https://github.com/AntennaPod/AntennaPod/blob/79ab68d56359dd8a02fd00441b8d2826f9c52c62/core/src/main/java/de/danoeh/antennapod/core/util/id3reader/ChapterReader.java) might be helpful.

**Comment by skynetua** [comment]
I was trying to do that few years ago with jaudiotagger library, it was problematic for m4a\m4b formats.

**Comment by Ph1b** [comment]
I have opened a ticket on the jaudiotagger library (a while ago)[https://bitbucket.org/ijabz/jaudiotagger/issues/64/chapter-marks] but their response was basically that we would have to implement this ourselves.

**Comment by skynetua** [comment]
We can try to do it. Take a look there: https://dl.dropboxusercontent.com/u/88537607/jaudiotagger_mod.zip
See getChapterMap() method in GenericAudioHeader.java and ChapterDataReader.java
We just need to merge it to current version and fix decompilation errors.

**Comment by Ph1b** [comment]
Where is that code coming from?

Another thing I invested was [this](https://github.com/wseemann/FFmpegMediaMetadataRetriever/issues/53).

**Comment by skynetua** [comment]
I don't remember, I can't find it in GP anymore. It was some player for audiobooks. I checked so many of them while investigating.

**Comment by Ph1b** [comment]
Im not sure about the license then. jaudiotagger is lgpl itself, but I won't just touch the code without knowing where it came from.

**Comment by skynetua** [comment]
FFmpeg looks pretty good however +14mb to app size can be significant.

**Comment by kevmitch** [comment]
isn't ffmpeg already being used? or is that not directly accessible by apps?

**Comment by forteller** [comment]
Can I just say that I really want this feature? :)

**Comment by wberrier** [comment]
I think this can be done with this library for m4b chapters:

 https://github.com/sannies/mp4parser

And they also have an example of getting meta data from mp4/m4b containers:

 https://github.com/sannies/mp4parser/blob/master/examples/src/main/java/org/mp4parser/examples/metadata/MetaDataRead.java

Atomic Parsley can be useful for showing the metadata tree, and may give some hints of where to get the chapter info from:

 AtomicParsley "m4b file" -T

I think it would also be nice to do chapter marks for webm/mka files, but I don't have any information about how to do that yet.

A comment about the issue description:

"Show chapters instead of files in BookPlay"  Does that mean that in my list of books it would show the chapters?  I have a 20 hour book with dozens and dozens of chapters.  That would really clutter up that list.  I would think that instead the "jump to" dialog would not only allow time input, but if the book has chapters, chapter selection input as well.

**Comment by rejhgadellaa** [comment]
+1 for m4b chapter support and /sub

**Comment by PaulWoitaschek** [comment]
I have a prototype ready which can read mp3 and MP4/m4a/m4b/aac.

However I have very little sample data to ensure it works for all kind of files. 
Can you send me files you know they have chapters to woitaschek@gmail.com ?

**Comment by PaulWoitaschek** [comment]
@wberrier 

Well the mp3 parsing turned out to be easy. However mp4 parsing is complicated. I got the auphonic chapter sample working: Auphonic Chapter Marks Demo, MP4 (M4A)

But there are other ways (like the chap atom) I don't understand. If someone wants to contribute on this that would be really great

```kotlin
  val isoFile = IsoFile("auphonic_chapters_demo.m4a")
  val appleBox: AppleLyricsBox = Path.getPath(isoFile, "/moov[0]/udta[0]/meta[0]/ilst/©lyr")
  val map = HashMap<Long, String>()
  appleBox.value.split("\n").filter { TIME_PATTERN.matcher(it).matches() }
    .forEach {
      val hours = it.substring(0, 2).toLong()
      val minutes = it.substring(3, 5).toLong()
      val seconds = it.substring(6, 8).toLong()
      val millis = it.substring(9, 12).toLong()

      val millisCummulative = TimeUnit.HOURS.toMillis(hours) +
        TimeUnit.MINUTES.toMillis(minutes) +
        TimeUnit.SECONDS.toMillis(seconds) +
        millis

      val text = it.substring(12, it.length)
        .replace(REPLACE_LINK, "")
        .trim(' ', '_', ':')

      map.put(millisCummulative, text)
    }
  println(map)
```
This parses the auphonic chapters correctly:
```
{
0=Intro, 22000=Sound analysis, 
34000=Adaptive leveler, 
60000=Audio restoration algorithms, 
76000=Output file formats, 
94000=External services, 
15000=Creating a new production, 
45000=Global loudness normalization, 
111500=Get a free account!
}
```

**Comment by wberrier** [comment]
Nice to see this coming along.  Here are a few librevox recordings that I "chapterized" in various formats (including m4b):

https://drive.google.com/drive/folders/0B5DxbhZHghuGMnRkY1VpakI2TjA?usp=sharing

**Comment by PaulWoitaschek** [comment]
@wberrier Got it working!

Keep sending more files, there are at least 4 different ways of how chapters are encoded in mp4 :D

**Comment by wberrier** [comment]
Nicely done!!

At least with audiobooks, this is by far the most popular chapter method for mp4.  I don't have any other chapter formats.

**Comment by PaulWoitaschek** [comment]
Okay, first snapshot is in beta + in the releases tab:

https://github.com/PaulWoitaschek/MaterialAudiobookPlayer/releases

If you have chapters you know are there but are not detected, please report!

**Comment by Eifelschaf** [comment]
Hi guys! There seem to be issues with the following file from a podcast: 

* Episode Download Link (138 MB): http://minkorrekt.de/podlove/file/1401/s/feed/c/m4a/methodisch_inkorrekt_Folge_93.m4a

There should be chapter marks in there as they usually have them.

**Comment by PaulWoitaschek** [comment]
Thanks!

I found a sample which parses the `chap` chapters. However I don't know C++, maybe someone can jump in:

Here: https://github.com/pcwalton/mp4v2/blob/master/src/mp4file.cpp#L2525

It calls `MP4TrackId chapterTrackId = FindChapterTrack();`. But how does it do that? I can only find a `FindChapterTrack` with a different signature. What are the arguments in that case?

**Comment by PaulWoitaschek** [comment]
Okay I got the chapter parsing working for all types of chapter marks.

I uploaded a new beta, please report if something is not working as expected.
Currently a few things are not implemented yet:

* The next button skips to the next file, not the next chapter
* The auto-rewind does not stop at a chapter mark and skips to the previous file

**Comment by Eifelschaf** [comment]
That looks good! Now one more thing for the UI would be a divider in the list to see where I've track ends and another one starts.

**Comment by wberrier** [comment]
@PaulWoitaschek this works great, nicely done!

I have a couple of files with issues...

In this one, the chapters don't show up at all:

http://media2.ldscdn.org/assets/scripture-and-lesson-support/teachings-of-presidents-of-the-church-gordon-b-hinckley/2017-01-00-teachings-of-presidents-of-the-church-gordon-b-hinckley-128k-eng.m4b?download=true

I have another file that I encoded myself, but for some reason, only the first 4 chapters show up.  On all the other files I've encoded myself they are correct.  Not sure what's up with that yet.  Here's a link to the file:

https://drive.google.com/open?id=0B5DxbhZHghuGNlg3TEZ2T1ZGVnM
